python/lsst/meas/extensions/shapeHSM/_hsm_higher_moments.py

```diff
@@ -0,0 +1,397 @@
+        default=False,
+    )
+
+
+@measBase.register("ext_shapeHSM_HigherOrderMomentsSource")
+class HigherOrderMomentsSourcePlugin(HigherOrderMomentsPlugin):
+    """Plugin for Higher Order Moments measurement of objects.
+
+    The moments are measured in normalized coordinates, where the normalized x
+    axis is along the major axis and the normalized y axis along the minor.
+    The moments are dependent only on the light profile, and does not scale
+    with the size or orientation of the object.
+
+    For any well-sampled image, the zeroth order moment is 1,
+    the first order moments are 0, and the second order moments are 0.5 for xx
+    and yy and 0 for xy. For a symmetric profile, the odd order moments are
+    zeros.
+
+    Notes
+    -----
+    This plugin requires the `ext_shapeHSM_HsmSourceMoments` plugin to be
+    enabled in order to measure the higher order moments, and raises a
+    FatalAlgorithmError otherwise. For accurate results, the weight function
+    used must match those used for first and second order moments. Hence, this
+    plugin does not use slots for centroids and shapes, but instead uses those
+    measured by the `ext_shapeHSM_HsmSourceMoments` explicitly.
+
+    The only known failure mode of this plugin is if
+    `ext_shapeHSM_HsmSourceMoments` measurement failed. The flags of that
+    plugin are informative here as well and should be used to filter out
+    unreliable measurements.
+    """
+
+    ConfigClass = HigherOrderMomentsSourceConfig
+
+    def __init__(self, config, name, schema, metadata, logName=None):
+        super().__init__(config, name, schema, metadata, logName=logName)
+
+        for suffix in self._generate_suffixes():
+            schema.addField(
+                schema.join(name, suffix),
+                type=float,
+                doc=f"Higher order moments M_{suffix} for source",
+            )
+
+    def measure(self, record, exposure):
+        # Docstring inherited.
+        M = np.zeros((2, 2))
+        try:
+            center = geom.Point2D(
+                record["ext_shapeHSM_HsmSourceMoments_x"],
+                record["ext_shapeHSM_HsmSourceMoments_y"],
+            )
+            M[0, 0] = record["ext_shapeHSM_HsmSourceMoments_xx"]
+            M[1, 1] = record["ext_shapeHSM_HsmSourceMoments_yy"]
+            M[0, 1] = M[1, 0] = record["ext_shapeHSM_HsmSourceMoments_xy"]
+        except KeyError:
+            raise measBase.FatalAlgorithmError("'ext_shapeHSM_HsmSourceMoments' plugin must be enabled.")
+
+        # Obtain the bounding box of the source footprint
+        bbox = record.getFootprint().getBBox()
+
+        bitValue = exposure.mask.getPlaneBitMask(self.config.badMaskPlanes)
+        badpix = (exposure.mask[bbox].array & bitValue) != 0
+
+        # Measure all the moments together to save time
+        try:
+            hm_measurement = self._calculate_higher_order_moments(
+                exposure.image[bbox],
```

Comment:
This needs to be a copy of the sub-image.

Reply:
Might be clearer if the copy is inside `_calculate_higher_order_moments`, but yeah, good catch.

Reply:
Copying inside `_calculate_higher_order_moments` would make unnecessary copies for `psfImage`. I fixed it to make a deep copy on a need-to basis.

Reply:
Alternatively, I could make the modification to the weight image and not have to make a deep copy ever.

Reply:
Right.  I just meant we don't want anyone to have to remember that `_calculate_higher_order_moments` modifies the image.  Modifying the weight is a great solution.

Reply:
The commit I just pushed looks to me like the cleanest solution - multiply the weight and image with no regards for the `badpix` and mask this product as configured. It avoids making expensive deep copies, and avoids any division of pixel values (as I currently have) which may become unstable if the pixel values are close to 0.